File: Java/chapter_8_challenge/app/src/main/java/com/glella/criminalintent/CrimeListFragment.java
package com.glella.criminalintent;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

public class CrimeListFragment extends Fragment {

    public RecyclerView mCrimesRecyclerView;
    private CrimeAdapter mAdapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_crime_list, container, false);

        mCrimesRecyclerView = (RecyclerView) view.findViewById(R.id.crime_recycler_view);
        mCrimesRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

        updateUI();

        return view;
    }

    private void updateUI() {
        CrimeLab crimeLab = CrimeLab.get(getActivity());
        List<Crime> crimes = crimeLab.getCrimes();

        mAdapter = new CrimeAdapter(crimes);
        mCrimesRecyclerView.setAdapter(mAdapter);
    }

    private class CrimeHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private TextView mTitleTextView;
        private TextView mDateTextView;
        private Crime mCrime;
        // Challenge
        private Button mContactPoliceButton;

        // Challenge - change signature
        //public CrimeHolder(LayoutInflater inflater, ViewGroup parent) {
        public CrimeHolder(LayoutInflater inflater, ViewGroup parent, int viewType) {
            // change signature of parent
            super(inflater.inflate(viewType, parent, false));
            itemView.setOnClickListener(this);

            mTitleTextView = (TextView) itemView.findViewById(R.id.crime_title);
            mDateTextView = (TextView) itemView.findViewById(R.id.crime_date);
        }

        public void bind(Crime crime) {
            mCrime = crime;
            mTitleTextView.setText(mCrime.getTitle());
            mDateTextView.setText(mCrime.getDate().toString());

            // Challenge for when the button is clicked
            if (mCrime.isRequiresPolice()) {
                mContactPoliceButton = (Button) itemView.findViewById(R.id.contact_police_button);
                mContactPoliceButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(getActivity(), "Police contacted for " + mCrime.getTitle(), Toast.LENGTH_SHORT).show();
                    }
                });
            }

        }

        @Override
        public void onClick(View view) {
            Toast.makeText(getActivity(), mCrime.getTitle() + " clicked!", Toast.LENGTH_SHORT).show();
        }
    }

    private class CrimeAdapter extends RecyclerView.Adapter<CrimeHolder> {

        private List<Crime> mCrimes;

        public CrimeAdapter(List<Crime> crimes) {
            mCrimes = crimes;
        }

        @NonNull
        @Override
        public CrimeHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {

            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
            // Challenge - include viewType
            //return new CrimeHolder(layoutInflater, viewGroup):
            return new CrimeHolder(layoutInflater, viewGroup, viewType);
        }

        @Override
        public void onBindViewHolder(@NonNull CrimeHolder holder, int position) {
            Crime crime = mCrimes.get(position);
            holder.bind(crime);
        }


        @Override
        public int getItemCount() {
            return mCrimes.size();
        }

        // Challenge
        @Override
        public int getItemViewType(int position) {
            if (mCrimes.get(position).isRequiresPolice()) {
                return R.layout.list_item_crime_police;
            } else {
                return R.layout.list_item_crime;
            }
        }
    }

}

File: Kotlin/chapter_8_challenge/app/src/main/java/com/glella/criminalintent/CrimeListFragment.kt
package com.glella.criminalintent

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast

class CrimeListFragment : Fragment() {


    lateinit var mCrimesRecyclerView: RecyclerView
    private var mAdapter: CrimeAdapter? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.fragment_crime_list, container, false)

        mCrimesRecyclerView = view.findViewById(R.id.crime_recycler_view) as RecyclerView
        mCrimesRecyclerView.setLayoutManager(LinearLayoutManager(activity))

        updateUI()

        return view
    }

    private fun updateUI() {
        val crimes = CrimeLab.mCrimes

        mAdapter = CrimeAdapter(crimes)
        mCrimesRecyclerView.setAdapter(mAdapter)
    }

    // Challenge - change signature
    //private inner class CrimeHolder(inflater: LayoutInflater, parent: ViewGroup) : RecyclerView.ViewHolder(inflater.inflate(R.layout.list_item_crime, parent, false)), View.OnClickListener {
    private inner class CrimeHolder(inflater: LayoutInflater, parent: ViewGroup, viewType: Int) : RecyclerView.ViewHolder(inflater.inflate(viewType, parent, false)), View.OnClickListener {

        private val mTitleTextView: TextView
        private val mDateTextView: TextView
        private var mCrime: Crime? = null
        // Challenge
        private var mContactPoliceButton: Button? = null

        init {
            itemView.setOnClickListener(this)

            mTitleTextView = itemView.findViewById(R.id.crime_title)
            mDateTextView = itemView.findViewById(R.id.crime_date)
        }

        fun bind(crime: Crime) {
            mCrime = crime
            mTitleTextView.setText(mCrime!!.mTitle)
            mDateTextView.setText(mCrime!!.mDate.toString())


            // Challenge for when the button is clicked
            if (crime.mRequiresPolice) {
                mContactPoliceButton = itemView.findViewById<View>(R.id.contact_police_button) as Button
                mContactPoliceButton?.setOnClickListener(View.OnClickListener { Toast.makeText(activity, "Police contacted for " + crime.mTitle, Toast.LENGTH_SHORT).show() })
            }
        }

        override fun onClick(view: View) {
            Toast.makeText(activity, mCrime!!.mTitle + " clicked!", Toast.LENGTH_SHORT).show()
        }
    }

    private inner class CrimeAdapter(private val mCrimes: List<Crime>) : RecyclerView.Adapter<CrimeHolder>() {


        override fun getItemCount(): Int {
            return mCrimes.size
        }

        override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): CrimeHolder {

            val layoutInflater = LayoutInflater.from(activity)
            // Challenge - include viewType
            //return CrimeHolder(layoutInflater, viewGroup)
            return CrimeHolder(layoutInflater, viewGroup, viewType)
        }

        override fun onBindViewHolder(holder: CrimeHolder, position: Int) {
            val crime = mCrimes[position]
            holder.bind(crime)
        }

        // Challenge
        override fun getItemViewType(position: Int): Int {
            return if (mCrimes[position].mRequiresPolice) {
                R.layout.list_item_crime_police
            } else {
                R.layout.list_item_crime
            }
        }
    }

}